include/mqtt/endpoint.hpp

```diff
@@ -8278,1217 +8319,3822 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
         case control_packet_type::suback:
             if (mqtt_connected_) {
-                ret = handle_suback(func);
+                process_suback(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::unsubscribe:
             if (mqtt_connected_) {
-                ret = handle_unsubscribe(func);
+                process_unsubscribe(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::unsuback:
             if (mqtt_connected_) {
-                ret = handle_unsuback(func);
+                process_unsuback(std::move(func), remaining_length_ < packet_bulk_read_limit_, std::move(self));
             }
             break;
         case control_packet_type::pingreq:
             if (mqtt_connected_) {
-                ret = handle_pingreq(func);
+                process_pingreq(std::move(func));
             }
             break;
         case control_packet_type::pingresp:
             if (mqtt_connected_) {
-                ret = handle_pingresp(func);
+                process_pingresp(std::move(func));
             }
             break;
         case control_packet_type::disconnect:
-            handle_disconnect(func);
-            ret = false;
+            process_disconnect(func, remaining_length_ < packet_bulk_read_limit_, std::move(self));
             break;
         case control_packet_type::auth:
-            ret = handle_auth(func);
+            process_auth(func, remaining_length_ < packet_bulk_read_limit_, std::move(self));
             break;
         default:
             break;
         }
-        if (ret) {
-            h_mqtt_message_processed_(std::move(func));
-        }
     }
 
-    bool handle_connect(async_handler_t const& func) {
-        std::size_t i = 0;
-        if (remaining_length_ < 10 || // *1
-            payload_[i++] != 0x00 ||
-            payload_[i++] != 0x04 ||
-            payload_[i++] != 'M' ||
-            payload_[i++] != 'Q' ||
-            payload_[i++] != 'T' ||
-            payload_[i++] != 'T') {
-            if (func) func(boost::system::errc::make_error_code(boost::system::errc::protocol_error));
-            return false;
-        }
-
-        auto version = static_cast<protocol_version>(payload_[i++]);
-        if (version != protocol_version::v3_1_1 && version != protocol_version::v5) {
-            if (func) {
-                func(boost::system::errc::make_error_code(boost::system::errc::protocol_not_supported));
-            }
-            return false;
-        }
-
-        if (version_ == protocol_version::undetermined) {
-            version_ = version;
+    // primitive read functions
+    void process_nbytes(
+        async_handler_t func,
+        buffer buf,
+        std::size_t size,
+        std::function<void(buffer, buffer, async_handler_t, this_type_sp)> handler,
```

Comment:
Would doing

```cpp
using process_handler_t = std::function<void(buffer, buffer, async_handler_t, this_type_sp)>;
```

improve the code?

Reply:
There are some variations as follows:

```cpp
        std::function<void(packet_id_t, buffer, async_handler_t, this_type_sp)> handler,
```

```cpp
        std::function<void(buffer, buffer, async_handler_t, this_type_sp)> handler,
```

So if we introduce type aliases, naming is difficult. I think that current implementation is good enough.

Reply:
Ok.